lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -1355,6 +1344,48 @@ def compute_timestamp(resource_type, record, time)
         ts_secs = timestamp.tv_sec
         ts_nanos = timestamp.tv_nsec
       end
+      ts_secs = begin
+                  Integer ts_secs
+                rescue ArgumentError, TypeError
+                  ts_secs
+                end
+      ts_nanos = begin
+                   Integer ts_nanos
+                 rescue ArgumentError, TypeError
+                   ts_nanos
+                 end
+
+      # Adjust timestamps from the future.
+      if ts_secs.is_a?(Integer) && ts_nanos.is_a?(Integer)
+        parsed_timestamp = Time.at(ts_secs + ts_nanos / 1_000_000_000.0)
+        end_of_year = Time.utc(Time.now.year, 12, 31, 23, 59, 59.999)
+        one_day_later = Time.now + 60 * 60 * 24
+        if parsed_timestamp > end_of_year
+          # This isn't the issue with system logs missing a year, because we are
+          # further than the current year. It is unlikely that users wanted logs
+          # more than a year in the future, This could be logs from a VM just
+          # starting up and the clock hasn't synced, or there was no reasonable
+          # time in the payload but we parsed some random number. Leave time as
+          # the default value so downstream backend will handle it.
+          ts_secs = 0
+          ts_nanos = 0
+        elsif parsed_timestamp > one_day_later
+          # This is the current year, but in the future. Most likely this is
+          # from some previous year. We assume it is last year. This could
+          # incorrectly label 2 year old logs, but they probably aren't super
+          # important and implies that fluentd is broken and resending unneeded
```

Comment:
> implies that fluentd is broken and resending unneeded logs.

Not clear. Clients could be ingesting old unrotated logs for the first time. There's no way for fluentd to know that these are old. I would remove this phrase.